infra/Pos/Communication/Relay/Logic.hs

```diff
@@ -75,249 +80,240 @@ type RelayWorkMode m =
-       , Message (InvOrData tag key contents)
-       , Relay m tag key contents
-       , MinRelayWorkMode m
+    :: forall m.
+       ( MinRelayWorkMode m
+       , MonadGStateCore m
+       )
+    => MempoolParams m
+    -> [(ListenerSpec m, OutSpecs)]
+handleMempoolL NoMempool = []
+handleMempoolL (KeyMempool tagP handleMempool) = pure $ listenerConv $ \__ourVerInfo ->
+  SizedCAHandler $ \__peerId conv ->
+      let handlingLoop = do
+              mbMsg <- fmap (withLimitedLength' $ convToSProxy conv) <$> recv conv
+              whenJust mbMsg $ \msg@MempoolMsg -> do
+                  let _ = msg `asProxyTypeOf` mmP
+                  res <- handleMempool
+                  case nonEmpty res of
+                      Nothing ->
+                          logDebug $ sformat
+                              ("We don't have mempool data "%shown) (typeRep tagP)
+                      Just xs -> do
+                          logDebug $ sformat ("We have mempool data "%shown) (typeRep tagP)
+                          mapM_ (send conv . InvMsg) xs
+                  handlingLoop
+       in handlingLoop
+  where
+    mmP = (const Proxy :: Proxy tag -> Proxy (MempoolMsg tag)) tagP
+
+handleDataOnlyL
+    :: forall contents m .
+       ( Bi (DataMsg contents)
+       , Message Void
+       , Message (DataMsg contents)
+       , Buildable contents
+       , RelayWorkMode m
        , MonadGStateCore m
+       , MessageLimited (DataMsg contents)
        )
-    => RelayProxy key tag contents
-    -> m (ListenerSpec m, OutSpecs)
-handleMempoolL proxy = reifyMsgLimit (Proxy @(MempoolMsg tag)) $
-  \(_ :: Proxy s) -> return $ listenerConv $
-    \_ __peerId ConversationActions{..} ->
-    whenJustM recv $ \(withLimitedLength @s -> msg@MempoolMsg{..}) -> do
-      let _ = mempoolCatchType proxy msg
-      processMessage () "Mempool" mmTag verifyMempoolTag $ do
-          res <- handleMempool mmTag
-          case nonEmpty res of
-              Nothing -> logDebug $ sformat
-                  ("We don't have mempool data "%build) mmTag
-              Just xs -> do
-                  logDebug $ sformat ("We have mempool data "%build) mmTag
-                  mapM_ (send . constructInvMsg mmTag) xs
+    => (contents -> m Bool)
+    -> (ListenerSpec m, OutSpecs)
+handleDataOnlyL handleData = listenerConv $ \__ourVerInfo ->
+  SizedCAHandler (\__peerId conv ->
+      let handlingLoop = do
+              mbMsg <- fmap (withLimitedLength' $ convToSProxy conv) <$> recv conv
+              whenJust mbMsg $ \DataMsg{..} -> do
+                  ifM (handleData dmContents)
+                      (propagateData $ DataOnlyPM dmContents)
+                      (logUseless dmContents)
+                  handlingLoop
+       in handlingLoop) :: SizedCAHandler Void (DataMsg contents) m
   where
-    constructInvMsg :: tag -> key -> InvOrData tag key contents
-    constructInvMsg tag = Left . InvMsg tag
+    logUseless dmContents = logDebug $ sformat
```

Comment:
I am not sure it should be debug. On the one hand, we can receive whatever data. On the other hand, if everybody is honest, it's unlikely that we will receive something bad. It can happen, yes, but rarely (I think).